File: YelpDataMining/StreamingAlgorithms/ikshita_mishra_task2.py
import math
from datetime import datetime, timezone
from pyspark import SparkContext
import json
import sys
import binascii
from pyspark.streaming import StreamingContext

sc = SparkContext()
sc.setLogLevel(logLevel="OFF")

hostname = "localhost"
portNum = int(sys.argv[1])
ssc = StreamingContext(sc, 5)
lines = ssc.socketTextStream(hostname, portNum)

words = lines.map(lambda x : (json.loads(x)['city']))


with open(sys.argv[2], 'w') as f:
    f.write("Time,Gound Truth,Estimation")
    f.write("\n")


def txt(x):
    with open(sys.argv[2], 'a') as f:
        f.write(str(x[0]) + "," + str(x[1]) + "," + str(x[2]))
        f.write("\n")


def flajolet(partition):

    lis = partition.collect()
    groundSet = set()
    sToInt = []
    for s in lis:
        sToInt.append(int(binascii.hexlify(s.encode('utf8')), 16))
        groundSet.add(s)
    hashFunc = [[87, 91, 671], [123, 192, 671], [35, 50, 671], [195, 164, 671],
                [136, 172, 671], [13, 19, 671], [93, 32, 671], [85, 95, 671], [19, 23, 671]]

    hashBit = [0] * len(hashFunc)

    m = len(sToInt)
    for num in range(0, len(hashFunc)):
        h = hashFunc[num]
        r = []
        for i in sToInt:
            hashVal = ((h[0] * i + h[1]) % h[2]) % m
            an = "{0:b}".format(hashVal)
            val = len(an) - len(an.rstrip('0'))
            if val != len(an):
                r.append(val)
            elif val == len(an):
                r.append(0)
        hashBit[num] = max(r)
    a1 = 0
    for i in range(0,3):
        a1 = a1 + 2 ** (hashBit[i])
    a1 = a1 / 3

    b1 = 0
    for j in range(3, 6):
        b1 = b1 + 2 ** (hashBit[j])
    b1 = b1 / 3

    c1 = 0
    for j in range(6, 9):
        c1 = c1 + 2 ** (hashBit[j])
    c1 = c1 / 3
    liM = sorted([a1, b1,c1])
    predictedCount = math.floor(liM[1])

    calTime = datetime.utcnow().replace(tzinfo=timezone.utc).astimezone(tz=None).strftime('%Y-%m-%d %H:%M:%S')
    a = (calTime,len(groundSet),predictedCount)
    txt(a)

words.window(30,10).foreachRDD(lambda rdd: flajolet(rdd))

ssc.start()
ssc.awaitTermination()
ssc.stop()

File: YelpDataMining/StreamingAlgorithms/ikshita_mishra_task2.scala
import org.apache.spark._
import org.apache.log4j.Logger
import org.apache.log4j.Level
import java.text.SimpleDateFormat
import java.util.{Date, TimeZone}
import org.json4s._
import java.io._
import breeze.linalg.max
import org.json4s.jackson.JsonMethods._
import org.apache.spark.streaming._

import scala.collection.mutable.ListBuffer

object ikshita_mishra_task2 {

  def main(args: Array[String]): Unit = {


    Logger.getLogger("org").setLevel(Level.OFF)
    Logger.getLogger("akka").setLevel(Level.OFF)


    val conf = new SparkConf().setMaster("local[*]").setAppName("Flajolet")
    val ssc = new StreamingContext(conf, Seconds(5))
    val lines = ssc.socketTextStream("localhost",args(0).toInt)

    val words = lines.map(x => pretty(render(parse(x)\"city")))


    val pw = new PrintWriter(new File(args(1)))
    pw.write("Time,Gound Truth,Estimation")
    pw.write("\n")
    pw.close()


    words.window(Seconds(30),Seconds(10)).foreachRDD(rdd => {
      val cities = rdd.collect()
      //println(cities)
      var groundSet = scala.collection.mutable.Set[Int]()
      val lis = cities.toList
      var sToInt: ListBuffer[Int] = scala.collection.mutable.ListBuffer.empty[Int]
      for ( i <- lis){
        val hashVal: Int = i.hashCode
        sToInt.append(hashVal)
        groundSet.add(hashVal)
      }


      val hashFunc: List[List[Int]] =
        List(List(87, 91, 671), List(123, 192, 671), List(35, 50,671),
          List(195, 164, 671), List(136, 172, 671), List(13, 19,671),
          List(93, 32, 671), List(85, 85, 671), List(19, 23,671))


      val hashBit: Array[Int] = Array.fill(hashFunc.length)(0)

      for (m <- hashFunc){
        var minLi: ListBuffer[Int] = scala.collection.mutable.ListBuffer.empty[Int]
        val num = hashFunc.indexOf(m)
        for (n <- sToInt)
        {

          val hashVal = ((m(0) * n + m(1)) % m(2)) % sToInt.length
          val an = hashVal.toBinaryString
          def rtrim(s: String) = s.replaceAll("0+$", "")
          val  zero = an.length - rtrim(an).length
          if (zero!=an.length){
            minLi.append(zero)
          }
          else if (zero==an.length){
            minLi.append(0)
          }
        }
        hashBit(num) = max(minLi)
      }
      var a1 = 0
      val aLi = List.range(0, 3)
      for (i <- aLi) {
        a1 = a1 + ((hashBit(i)) * (hashBit(i)))
      }
      a1 = a1 / aLi.length

      var b1 = 0
      val bLi = List.range(3, 6)
      for (j <- bLi) {
        b1 = b1 + ((hashBit(j)) * (hashBit(j)))
      }
      b1 = b1 / bLi.length

      var c1 = 0
      val cLi = List.range(6, 9)
      for (k <- cLi) {
        c1 = c1 + ((hashBit(k)) * (hashBit(k)))
      }
      c1 = c1 / cLi.length

      val li = List(a1, b1, c1).sorted
      val predictedCount = (li(1)).toInt

      val dateFor = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss")
      dateFor.setTimeZone(TimeZone.getTimeZone("America/Los_Angeles"))
      var subD = new Date()
      val dt = dateFor.format(subD)
      val fw = new FileWriter(args(1), true)
      fw.write(dt.toString() +","+ groundSet.size.toString()+","+ predictedCount.toString()+"\n")
      fw.close()

    })

    ssc.start()
    ssc.awaitTermination()

  }

}